AK+Time: Less chrashy, again

BenWiederhake opened the pull request:
This PR:
- Makes `AK/Time.h` overflow-correct again. Keep in mind that signed overflow is UB, and we probably don't want UB in the Kernel, do we?
- Makes `AK/Time.h` crash less in the presence of invalid inputs. We probably don't want to crash the Kernel if someone inserts a CD with an invalid month, do we?
- Tests for all of the above things to a large extent

Note that this PR builds on top of #17842. Since this PR goes far beyond the scope of #17842, and because #17842 has already been delayed more than enough, I don't want to pile on top of it. Let's first merge #17842, and then this PR.

Or as @alimpfard put it:
> Does your code contain unexpected integer overflow? of course it does! contact BenW to find out why!
– https://benwiederhake.github.io/freenode-serenity-archive/quote-132827.html

Note that getting AK::Time correct seems to be very hard, as I have a history of fixing overflow-bugs around this code: #5323 #5740 

Also note that *callers* into AK::Time often try to do their own conversions beforehand, which often introduces even more signed overflow UB, but I didn't address that in this PR.

This PR is "draft" until #17842 is merged. CC @kleinesfilmroellchen in case you're interested.

kleinesfilmroellchen commented on Kernel/FileSystem/ISO9660FS/Inode.cpp:
```diff
@@ -194,6 +194,8 @@ UnixDateTime ISO9660Inode::parse_numerical_date_time(ISO::NumericalDateAndTime c
 
     // FIXME: This ignores timezone information in date.
+    // FIXME: This calls `VERIFY(month >= 1 && month <= 12)` (but irritatingly allows days 0 and 255),
+    // which means we can crash the kernel by inserting a CD with an invalid recording month.
```
Interesting, since I only moved the from_parts parsing code from one class to another.

BenWiederhake replied:
I'm not accusing you of introducing any bugs, sorry :D

This `VERIFY` seems to have existed for a very long time, at least since 81079ae61636c1f7f32f64466d55d82ae72d67d3. Now that this code is being called in the Kernel on unsanitized data, it *definitely* must not crash. I'm not sure how long this has already been the case, it's just that while reading your PR I noticed this oddity.

kleinesfilmroellchen commented on Tests/AK/TestDuration.cpp:
```diff
@@ -328,6 +328,43 @@ BENCHMARK_CASE(years_to_days_since_epoch_benchmark)
+    // FIXME shouldn't crash: EXPECT_EQ(days_since_epoch(1971, 0, 0), 333);
+    // FIXME shouldn't crash: EXPECT_EQ(days_since_epoch(1971, 13, 3), ???);
+
+    // I can't easily verify that these values are perfectly exact and correct, but they're close enough.
```
(1) Unix time is always Gregorian even before that calendar was really introduced, so we don't have to observe the Julian/Gregorian calendar discontinuity in 1500-something.
(2) Unix time doesn't know that year 0 doesn't exist, so all negative years are shifted by 1
(3) Unix time doesn't know that leap year rules change before the Gregorian/Julian change

... and the list goes on. This API should really be removed at some point since it's hopelessly naive without precise timezone and calendar information, I only left it in to not overburden the initial PR. We should really be using special parsers depending on where the data comes from and what calendar they define; for instance the Kernel usage is limited to the CD-ROM definition of a Gregorian/Julian calendar with Unix timezone.

kleinesfilmroellchen commented on Tests/AK/TestDuration.cpp:
```diff
@@ -490,3 +527,95 @@ TEST_CASE(user_defined_literals)
+    EXPECT_DURATION(UnixDateTime::from_unix_time_parts(1972, 2, 29, 0, 0, 0, 0).offset_to_epoch(), 68169600, 0);
+    EXPECT_DURATION(UnixDateTime::from_unix_time_parts(1972, 3, 1, 0, 0, 0, 0).offset_to_epoch(), 68256000, 0);
+
+    EXPECT_DURATION(UnixDateTime::from_unix_time_parts(2023, 5, 24, 18, 44, 40, 0).offset_to_epoch(), 1684953880, 0);
```
\<sidenote>I hope this doesn't make Serenity look horribly dated when someone looks at this code in 30 years, just like I cringe at Java Time API example code that references dates from 2001\</sidenote>

kleinesfilmroellchen commented on Tests/AK/TestDuration.cpp:
```diff
@@ -565,39 +565,28 @@ TEST_CASE(from_unix_time_parts_negative)
 {
     // Negative "common" values. These aren't really that well-defined, but we must make sure we don't crash.
     // FIXME: We shouldn't VERIFY(), especially because this is used in the Kernel!
-    dbgln("line {}", __LINE__);
```
probably should not have been committed in the first place

BenWiederhake replied:
Absolutely

kleinesfilmroellchen commented on AK/Time.h:
```diff
@@ -54,7 +54,10 @@ unsigned day_of_week(int year, unsigned month, int day);
 {
-    VERIFY(month >= 1 && month <= 12);
+    if (is_constant_evaluated())
+        VERIFY(month >= 1 && month <= 12); // Note that this fails to compile if and only if month is invalid.
```
"fails to compile" -> "fails" (since constexpr is only optionally constant evaluated, as opposed to consteval)

BenWiederhake replied:
No, it doesn't fail. That's the weird thing I tried to document: We cannot `outln` in constexpr code. However, for some reason this only fails to compile if and only if `month >= 1 && month <= 12` evaluates to false. But it is never executed either way.

What's a better way to document this?

kleinesfilmroellchen reviewed (changes requested):
As said in one of the comments, the from_parts API should probably be removed as soon as possible. I've put more details in a new issue: #19030

BenWiederhake commented:
Changes since last push:
- Rebased on master, now that #17842 is merged
- Removed debug lines that accidentally made it into the draft
- Clarified a comment about the `constexpr VERIFY()` – thanks @kleinesfilmroellchen!

I agree that removing the from_parts API would be nice, but any replacement would still need to have a lot of the same code, so I don't think this is reasonably possible. After all, we probably don't care about "correct" date parsing of CDROMs that claim to have been made in 1400.
It's much more important that we don't crash the Kernel in that case.
